File: problems/0125-valid-palindrome/go/solution.go
package solution

import "unicode"

// Valid Palindrome — https://leetcode.com/problems/valid-palindrome/
func isPalindrome(s string) bool {
	r := []rune(s)
	i, j := 0, len(r)-1
	for i < j {
		for i < j && !isAlnum(r[i]) {
			i++
		}
		for i < j && !isAlnum(r[j]) {
			j--
		}
		if unicode.ToLower(r[i]) != unicode.ToLower(r[j]) {
			return false
		}
		i++
		j--
	}
	return true
}

func isAlnum(r rune) bool {
	return unicode.IsLetter(r) || unicode.IsDigit(r)
}

File: problems/0125-valid-palindrome/typescript/solution.ts
// Valid Palindrome — https://leetcode.com/problems/valid-palindrome/
export function isPalindrome(s: string): boolean {
  const isAlnum = (c: string): boolean => /[a-z0-9]/i.test(c);
  let i = 0;
  let j = s.length - 1;
  while (i < j) {
    while (i < j && !isAlnum(s[i])) i++;
    while (i < j && !isAlnum(s[j])) j--;
    if (s[i].toLowerCase() !== s[j].toLowerCase()) return false;
    i++;
    j--;
  }
  return true;
}
